C/plugins/storage/common/disk_monitor.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Fledge storage service.
+ *
+ * Copyright (c) 2024 Dianomic Systems
+ *
+ * Released under the Apache 2.0 Licence
+ *
+ * Author: Mark Riddoch
+ */
+#include <disk_monitor.h>
+#include <logger.h>
+#include <sys/vfs.h>
+#include <string.h>
+
+using namespace std;
+
+/**
+ * Construt a disk space monitor class
+ *
+ * It monitors the free space on two path, since the storage service may use
+ * different locations for reading and configuration storage. If they are both
+ * the same file system then the monitor is only done once.
+ *
+ * If the free space falls below 5% a fatal error is written to the error log
+ * If it falls below 10% a warning is written advising that disk space should be released
+ * It attempts to predict when storage will bebeome exhausted. If it is less then 14 days
+ * it will report this to the error log once a day.
+ * If it is less than 72 hours it will report this once per hour.
+ *
+ * All reporting is to the system log
+ */
+DiskSpaceMonitor::DiskSpaceMonitor(const string& path1, const string& path2) :
+	m_dbPath1(path1), m_dbPath2(path2), m_started(false), m_sameDevice(false), m_lastCheck(0),
+	m_lastPerc1(0.0), m_lastPerc2(0.0), m_lastPrediction1(0.0), m_lastPrediction2(0.0)
+{
+	m_logger = Logger::getLogger();
+}
+
+/**
+ * Called periodically to monitor the disk usage
+ *
+ * @param interval	The number of seconds between calls
+ */
+void DiskSpaceMonitor::periodic(int interval)
+{
+	struct statfs stf1, stf2;
+
+	if (!m_started)
+	{
+		if (statfs(m_dbPath1.c_str(), &stf1) != 0)
+		{
+			m_logger->error("Can;t stafs %s", m_dbPath1.c_str());
+			return;
+		}
+		if (statfs(m_dbPath2.c_str(), &stf2) != 0)
+		{
+			m_logger->error("Can;t stafs %s", m_dbPath2.c_str());
+			return;
+		}
+		if (memcmp(&stf1.f_fsid, &stf2.f_fsid, sizeof(fsid_t)) == 0)	// Same filesystem
+		{
+			m_sameDevice = true;
+		}
+		m_started = true;
+	}
+	m_lastCheck += interval;
+
+	if (m_lastCheck < CHECK_THRESHOLD)
+	{
+		// Do not check too frerquenctly
+		return;
+	}
+	m_lastCheck = 0;
+
+	
+	if (statfs(m_dbPath1.c_str(), &stf1) == 0)
+	{
+		unsigned long freespace = (unsigned long)stf1.f_bavail;
+		unsigned long totalspace = (unsigned long)stf1.f_blocks;
+
+		double perc = (double)(freespace  * 100.0) / totalspace;
+
+		if (perc < 5.0)
+		{
+			m_logger->fatal("Disk space is critically low. Urgent action required, continuing may result in data corruption");
+		}
+		else if (perc < 10.0)
+		{
+			m_logger->error("Available disk space is becoming low, please consider releasing more disk space");
+		}
+		if (m_lastPerc1 > 0.0)
+		{
+			double diff = m_lastPerc1 - perc;
```

Comment:
diff over a very small duration like 15 seconds may not be representative of speed of disk filling. Wait a bit longer?